### PR TITLE
fix precisions & add bit support

### PR DESCRIPTION
**Pull request by BugenZhao**
### What problem does this PR solve?

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- PR to update `pingcap/tidb-ansible`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
Please add a release note.
If you don't think this PR needs a release note then fill it with None.
```

### CODE REVIEW

File: components/backup_text/src/hr_value.rs [hunk 1]
```diff
@@ -10,6 +10,7 @@ pub struct RowV1 {
     pub ids: Vec<u32>,
     #[serde(rename = "d")]
     pub datums: Vec<HrDatum>,
+    // pub raw: Vec<u8>,
```
Comment (NingLin-P) [inline components/backup_text/src/hr_value.rs]: Needs clean-up.

File: components/tidb_query_datatype/src/expr/ctx.rs [hunk 2]
```diff
@@ -59,12 +59,23 @@ impl SqlMode {
...
+
 const DEFAULT_MAX_WARNING_CNT: usize = 64;
 
 #[derive(Clone, Debug)]
 pub struct EvalConfig {
     /// timezone to use when parse/calculate time.
     pub tz: Tz,
+    /// flags used for datum encoding
+    pub encoding_flag: EncodingFlag,
     pub flag: Flag,
```
Comment (NingLin-P) [inline components/tidb_query_datatype/src/expr/ctx.rs:79]: Maybe better to add the `encoding_flag` into `flag`

File: components/tidb_query_datatype/src/codec/mysql/decimal.rs [hunk 3]
```diff
@@ -1007,7 +1022,7 @@ impl Decimal {
     }
 
     /// Get the least precision and fraction count to encode this decimal completely.
-    pub fn prec_and_frac(&self) -> (u8, u8) {
+    pub fn least_prec_and_frac(&self) -> (u8, u8) {
```
Comment (NingLin-P) [inline components/tidb_query_datatype/src/codec/mysql/decimal.rs]: How about keep this function name unchanged

File: components/tidb_query_datatype/src/codec/mysql/decimal.rs [hunk 4]
```diff
@@ -907,6 +918,9 @@ pub struct Decimal {
     /// An array of u32 words.
     /// A word is an u32 value can hold 9 digits.(0 <= word < wordBase)
     word_buf: [u32; 9],
+
+    /// Extra fields that only exists in TiKV
+    extra: DecimalExtra,
```
Comment (NingLin-P) [inline components/tidb_query_datatype/src/codec/mysql/decimal.rs]: How about just `raw_int_cnt: u8` or `raw_prec: u8`

File: components/tidb_query_datatype/src/expr/ctx.rs (hunk 2 above)
Comment (BugenZhao) [reply components/tidb_query_datatype/src/expr/ctx.rs:79]: It seems that `Flag` struct matches the `flags` field from pb struct `DagRequest`. Thus I'm not sure whether it is safe to add it.

File: components/tidb_query_datatype/src/codec/mysql/decimal.rs (hunk 4 above)
Comment (BugenZhao) [reply components/tidb_query_datatype/src/codec/mysql/decimal.rs]: Considering that we may construct a `Decimal` from a byte slice in `read_decimal_from_chunk`, encapsulating all other fields that are not in TiDB struct may help us better init theses remaining fields manually.
Also, in this way we can also ensure the correct encoding size by
```rust
const_assert_eq!(
    DECIMAL_STRUCT_SIZE,
    mem::size_of::<Decimal>() - mem::size_of::<DecimalExtra>()
);
```

File: components/backup_text/src/sst_to_text.rs [hunk 5]
```diff
@@ -11,25 +11,23 @@ use crate::{
...
 pub fn kv_to_text(key: &[u8], val: &[u8], table: &TableInfo) -> CodecResult<String> {
-    let ctx = &mut EvalContext::default();
+    let ctx = &mut eval_context();
     let columns_info = table.get_columns();
     let column_id_info: HashMap<i64, &'_ ColumnInfo, _> = columns_info
         .iter()
         .map(|ci| (ci.get_column_id(), ci))
         .collect();
 
+    // let raw = val.to_vec();
```
Comment (NingLin-P) [inline components/backup_text/src/sst_to_text.rs]: ditto

**Pull request merged**